homeassistant/components/feedreader/coordinator.py

```diff
@@ -0,0 +1,247 @@
+"""Data update coordinator for RSS/Atom feeds."""
+
+from __future__ import annotations
+
+from calendar import timegm
+from datetime import datetime, timedelta
+from logging import getLogger
+import os
+import pickle
+from time import gmtime, struct_time
+
+import feedparser
+
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.storage import Store
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+from homeassistant.util import dt as dt_util
+
+from .const import DELAY_SAVE, DOMAIN, EVENT_FEEDREADER, STORAGE_VERSION
+
+_LOGGER = getLogger(__name__)
+
+
+class FeedReaderCoordinator(DataUpdateCoordinator):
+    """Abstraction over Feedparser module."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        url: str,
+        scan_interval: timedelta,
+        max_entries: int,
+        storage: StoredData,
+    ) -> None:
+        """Initialize the FeedManager object, poll as per scan interval."""
+        super().__init__(
+            hass=hass,
+            logger=_LOGGER,
+            name=f"{DOMAIN} {url}",
+            update_interval=scan_interval,
+        )
+        self._url = url
+        self._max_entries = max_entries
+        self._feed: feedparser.FeedParserDict | None = None
+        self._firstrun = True
+        self._storage = storage
+        self._last_entry_timestamp: struct_time | None = None
+        self._has_published_parsed = False
+        self._has_updated_parsed = False
+        self._event_type = EVENT_FEEDREADER
+        self._feed_id = url
+        _LOGGER.debug(
+            "coordinator initi with url:%s scan_interval:%s", url, scan_interval
+        )
+
+    def _log_no_entries(self) -> None:
+        """Send no entries log at debug level."""
+        _LOGGER.debug("No new entries to be published in feed %s", self._url)
+
+    async def _async_update_data(self) -> dict:
+        """Update the feed and publish new entries to the event bus."""
+        last_entry_timestamp = await self.hass.async_add_executor_job(self._update)
+        if last_entry_timestamp:
+            self._storage.async_put_timestamp(self._feed_id, last_entry_timestamp)
+        return {}
+
+    def _update(self) -> struct_time | None:
```

Comment:
Move all non-I/O operations into the `async_update_data` method

Reply:
b5a5211a048ca16cc4355fa2d3a6a87aeae67385